src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java

```diff
@@ -0,0 +1,507 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.ibm.cloud.sdk.core.test.security;
+
+import static com.ibm.cloud.sdk.core.test.TestUtils.loadFixture;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PowerMockIgnore;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import com.ibm.cloud.sdk.core.security.Authenticator;
+import com.ibm.cloud.sdk.core.security.CloudPakForDataAuthenticator;
+import com.ibm.cloud.sdk.core.security.CloudPakForDataServiceInstanceAuthenticator;
+import com.ibm.cloud.sdk.core.security.Cp4dServiceInstanceTokenResponse;
+import com.ibm.cloud.sdk.core.service.exception.ServiceResponseException;
+import com.ibm.cloud.sdk.core.test.BaseServiceUnitTest;
+import com.ibm.cloud.sdk.core.util.Clock;
+
+import okhttp3.Headers;
+import okhttp3.Request;
+import okhttp3.mockwebserver.RecordedRequest;
+
+@PrepareForTest({ Clock.class })
+@PowerMockIgnore({
+    "javax.net.ssl.*",
+    "okhttp3.*",
+    "okio.*"
+})
+public class Cp4dServiceInstanceAuthenticatorTest extends BaseServiceUnitTest {
+
+  // Token with issued-at time of 1574353085 and expiration time of 1574453956
+  private Cp4dServiceInstanceTokenResponse tokenData;
+
+  // Token with issued-at time of 1602788645 and expiration time of 1999999999
+  private Cp4dServiceInstanceTokenResponse refreshedTokenData;
+
+  private String url;
+  private String testUsername = "test-username";
+  private String testApikey = "test-apikey";
+  private String testServiceInstanceId = "test-serviceInstanceId";
+
+  @Override
+  @BeforeMethod
+  public void setUp() throws Exception {
+    super.setUp();
+    url = getMockWebServerUrl();
+    tokenData = loadFixture("src/test/resources/cp4d_service_instance_token.json",
+        Cp4dServiceInstanceTokenResponse.class);
+    refreshedTokenData = loadFixture("src/test/resources/refreshed_cp4d_service_instance_token.json",
+        Cp4dServiceInstanceTokenResponse.class);
+  }
+
+
+  //
+  // Tests involving the new Builder class and fromConfiguration() method.
```

Comment:
```suggestion
  // Tests involving the Builder class and fromConfiguration() method.
```
The whole thing is new at this point :)